cmd/ftl/cmd_dev.go

```diff
@@ -0,0 +1,259 @@
+		dir := filepath.Dir(toml)
+		if _, ok := d.modules[dir]; !ok {
+			d.modules[dir] = moduleFolderInfo{
+				LastModTime: time.Now(),
+			}
+		}
+	}
+
+	for dir := range d.modules {
+		found := false
+		for _, toml := range tomls {
+			if filepath.Dir(toml) == dir {
+				found = true
+				break
+			}
+		}
+		if !found {
+			delete(d.modules, dir) // Remove deleted module from d.modules
+		}
+	}
+}
+
+func (d *devCmd) updateFileInfo(dir string) error {
+	config, err := moduleconfig.LoadConfig(dir)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	ignores := loadGitIgnore(os.DirFS(dir), ".")
+	d.modules[dir] = moduleFolderInfo{}
+
+	err = walkDir(dir, ignores, func(srcPath string, entry fs.DirEntry) error {
+		for _, pattern := range config.Watch {
+			relativePath, err := filepath.Rel(dir, srcPath)
+			if err != nil {
+				return errors.WithStack(err)
+			}
+
+			match, err := doublestar.PathMatch(pattern, relativePath)
+			if err != nil {
+				return errors.WithStack(err)
+			}
+
+			if match && !entry.IsDir() {
+				fileInfo, err := entry.Info()
+				if err != nil {
+					return errors.WithStack(err)
+				}
+
+				module := d.modules[dir]
+				module.NumFiles++
+				if fileInfo.ModTime().After(module.LastModTime) {
+					module.LastModTime = fileInfo.ModTime()
+				}
+				d.modules[dir] = module
+			}
+		}
+
+		return nil
+	})
+
+	return err
+}
+
+// errSkip is returned by walkDir to skip a file or directory.
+var errSkip = errors.New("skip directory")
+
+// Depth-first walk of dir executing fn after each entry.
+func walkDir(dir string, ignores []string, fn func(path string, d fs.DirEntry) error) error {
+	dirInfo, err := os.Stat(dir)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	if err = fn(dir, fs.FileInfoToDirEntry(dirInfo)); err != nil {
+		if errors.Is(err, errSkip) {
+			return nil
+		}
+		return errors.WithStack(err)
+	}
+	entries, err := os.ReadDir(dir)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	var dirs []os.DirEntry
+
+	// Process files first, then recurse into directories.
+	for _, entry := range entries {
+		fullPath := filepath.Join(dir, entry.Name())
+
+		// Check if the path matches any ignore pattern
+		shouldIgnore := false
+		for _, pattern := range ignores {
+			match, err := doublestar.PathMatch(pattern, fullPath)
+			if err != nil {
+				return errors.WithStack(err)
+			}
+			if match {
+				shouldIgnore = true
+				break
+			}
+		}
+
+		if shouldIgnore {
+			continue // Skip this entry
+		}
+
+		if entry.IsDir() {
+			dirs = append(dirs, entry)
+		} else {
+			if err = fn(fullPath, entry); err != nil {
+				if errors.Is(err, errSkip) {
+					// If errSkip is found in a file, skip the remaining files in this directory
+					return nil
+				}
+				return errors.WithStack(err)
+			}
+		}
+	}
+
+	// Then, recurse into subdirectories
+	for _, dirEntry := range dirs {
+		dirPath := filepath.Join(dir, dirEntry.Name())
+		ignores := loadGitIgnore(os.DirFS(dirPath), ".")
```

Comment:
I think you want to append to the existing ignores here rather than replacing? And you'll need to prepend the current directory to all of the paths returned by this, I think.

Reply:
Ahh good call! Thanks :)

I think the prepending might be ok since I use the relative path here, but I could be wrong.

```go
			relativePath, err := filepath.Rel(dir, srcPath)
			if err != nil {
				return errors.WithStack(err)
			}

			match, err := doublestar.PathMatch(pattern, relativePath)
```